### PR TITLE
feat: non-inclusion circuits

### PR DESCRIPTION
**Pull request by sergeytimoshin**

### CODE REVIEW

**Comment by SwenSchaeferjohann** [comment]
@sergeytimoshin need to rebase

**Comment by SwenSchaeferjohann** [comment]
looks cool, ccing @ananas-block

File: circuit-lib/circuit-lib.circom/src/non-inclusion-merkle-proof/NonInclusionMerkleProof.circom [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+        // Inclusion Proof Attempt
+        component merkleProof = MerkleProof(levels);
+        merkleProof.leaf <== leaf;
+        merkleProof.pathElements <== merkleProofHashedIndexedElementLeaf;
+        merkleProof.pathIndices <== indexHashedIndexedElementLeaf;
+
+        log("NonInclusionProof, root = ", root, ", merkleProof.root = ", merkleProof.root);
+
+        merkleProof.root === root;
```
Comment (ananas-block) [inline circuit-lib/circuit-lib.circom/src/non-inclusion-merkle-proof/NonInclusionMerkleProof.circom:108]: Is there a reason to keep the logs?

File: circuit-lib/circuit-lib.circom/src/non-inclusion-merkle-proof/NonInclusionMerkleProof.circom [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        component lessThanHigher = LessThan(252);
+        lessThanHigher.in[0] <== value;
+        lessThanHigher.in[1] <== leafHigherRangeValue;
+        signal notIncludedValueLessThanLeafHigherRangeValue <== lessThanHigher.out;
+        notIncludedValueLessThanLeafHigherRangeValue === 1;
+
+        
+        log("NonInclusionProof, calculating leaf...");
+        // Leaf Calculation
+        component poseidon = Poseidon(3); // Poseidon with 3 inputs
```
Comment (ananas-block) [inline circuit-lib/circuit-lib.circom/src/non-inclusion-merkle-proof/NonInclusionMerkleProof.circom:92]:
```suggestion
        component poseidon = Poseidon(3);
```

File: circuit-lib/circuit-lib.circom/src/non-inclusion-merkle-proof/NonInclusionMerkleProof.circom [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        component higherThanLower = LessThan(252);
+        higherThanLower.in[0] <== leafLowerRangeValue;
+        higherThanLower.in[1] <== value;
+        signal leafLowerRangeValueLessThanNotIncludedValue <== higherThanLower.out;
+        leafLowerRangeValueLessThanNotIncludedValue === 1;
+
+
+        log("NonInclusionProof, checking lessThanHigher...");
+        // check that notIncludedValue less than leafHigherRangeValue
+        // TODO: check that value is in Fr(254) on-chain
```
Comment (ananas-block) [inline circuit-lib/circuit-lib.circom/src/non-inclusion-merkle-proof/NonInclusionMerkleProof.circom:82]: what was the error here?

File: circuit-lib/circuit-lib.js/tests/merkle-tree.test.ts [hunk 4]
```diff
@@ -1,41 +1,38 @@
...
-  before(async () => {
-    lightWasm = await WasmFactory.getInstance();
-  });
-
-  it("merkle proofgen", async () => {
+  it("inclusion merkle proof", async () => {
     const hasher = await WasmFactory.getInstance();
     const merkleHeights = [26];
-    const utxos = [1, 2, 3, 4, 8];
+    const utxos = [1]; //, 2, 3, 4, 8];
```
Comment (ananas-block) [inline circuit-lib/circuit-lib.js/tests/merkle-tree.test.ts:30]: is this commented on purpose?

File: circuit-lib/circuit-lib.js/tests/merkle-tree.test.ts [hunk 5]
```diff
@@ -49,12 +46,13 @@ describe("Tests", () => {
         };
 
         const inputs_json = JSON.stringify(inputs);
+        console.log("inclusion, inputs: ", inputs_json);
```
Comment (ananas-block) [inline circuit-lib/circuit-lib.js/tests/merkle-tree.test.ts:49]: leftover log

File: circuit-lib/circuitlib-rs/src/verifying_keys/combined_26_1_1/mod.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
+use groth16_solana::groth16::Groth16Verifyingkey;
+
+pub const VERIFYINGKEY: Groth16Verifyingkey = Groth16Verifyingkey {
+    nr_pubinputs: 5,
```
Comment (ananas-block) [inline circuit-lib/circuitlib-rs/src/verifying_keys/combined_26_1_1/mod.rs:4]: The number of public inputs is one too high, can you change that in the script that generates these files?
(You can just decrease it by 1(vk_ic.length-1).)

File: circuit-lib/circuit-lib.circom/scripts/buildCombinedMerkleCircuits.sh [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+#     execute_commands "combined" "$MERKLE_TREE_HEIGHT" "$i" "$POWERS_OF_TAU" 1 || exit
+# done
+
+# execute_commands "combined" "$MERKLE_TREE_HEIGHT" 8 "$POWERS_OF_TAU" 1 || exit
+
+# for ((i=1; i<=MAX_COUNT; i++)); do
+#     execute_commands "combined" "$MERKLE_TREE_HEIGHT" "$i" "$POWERS_OF_TAU" 2 || exit
+# done
+
+# execute_commands "combined" "$MERKLE_TREE_HEIGHT" 8 "$POWERS_OF_TAU" 2 || exit
```
Comment (ananas-block) [inline circuit-lib/circuit-lib.circom/scripts/buildCombinedMerkleCircuits.sh:118]: lots of commented code here

File: merkle-tree/indexed/src/array.rs [hunk 8]
```diff
@@ -128,6 +129,16 @@ where
...
+
+        let hashBN = BigInt::from_bytes_be(Sign::Plus, &hash);
+
+        println!(
+            "hash={:?}, value={:?}, next_index={:?}, next_value={:?}",
+            hashBN,
+            self.value.to_bytes_be(),
+            self.next_index.to_be_bytes(),
+            next_value
+        );
```
Comment (ananas-block) [inline merkle-tree/indexed/src/array.rs:141]: debug print

**Review by ananas-block — COMMENTED** [review]
Looks good in general, lots of leftover comments and prints.

**Pull request closed without merging**